Template Haskell: ByteCodeLink.lookupCE During interactive linking, GHCi couldn't find the following symbol

The new Template Haskell support seems to get further at work, now failing with:

```
ghcide version: 0.0.5 (GHC: 8.8.1)
Ghcide setup tester in /home/ollie/work/circuithub.
Report bugs at https://github.com/digital-asset/ghcide/issues

Step 1/6: Finding files to test in /home/ollie/work/circuithub
Found 1 files

Step 2/6: Looking for hie.yaml files that control setup
Found 1 cradle

Step 3/6, Cradle 1/1: Loading ./hie.yaml

Step 4/6, Cradle 1/1: Loading GHC Session

Step 5/6: Initializing the IDE

Step 6/6: Type checking the files
ghcide: ^^ Could not load 'filepickerzmpolicyzm0zi2zi6zmLUnJDYkanFc1p7RlDh5zzSr_NetworkziFilepickerziHandle_FilepickerHandle_closure', dependency unresolved. See top entry above.

File:     entities/Model/Common.hs
Range:    1:0-100001:0
Source:   typecheck
Severity: DsError
Message: 
  Program error:
  ByteCodeLink.lookupCE
  During interactive linking, GHCi couldn't find the following symbol:
  filepickerzmpolicyzm0zi2zi6zmLUnJDYkanFc1p7RlDh5zzSr_NetworkziFilepickerziHandle_FilepickerHandle_closure
  This may be due to you not asking GHCi to load extra object files,
  archives or DLLs needed by your current session. Restart GHCi, specifying
  the missing library using the -L/path/to/object/dir and -lmissinglibname
  flags, or simply by naming the relevant files on the GHCi command line.
  Alternatively, this link failure might indicate a bug in GHCi.
  If you suspect the latter, please send a bug report to:
  glasgow-haskell-bugs@haskell.org
ghcide: ^^ Could not load 'OctopartV3ziTypesziCore_Uid_closure', dependency unresolved. See top entry above.

File:     entities/Model/OctopartV3.hs
Range:    1:0-100001:0
Source:   typecheck
Severity: DsError
Message: 
  Program error:
  ByteCodeLink.lookupCE
  During interactive linking, GHCi couldn't find the following symbol:
  OctopartV3ziTypesziCore_Uid_closure
  This may be due to you not asking GHCi to load extra object files,
  archives or DLLs needed by your current session. Restart GHCi, specifying
  the missing library using the -L/path/to/object/dir and -lmissinglibname
  flags, or simply by naming the relevant files on the GHCi command line.
  Alternatively, this link failure might indicate a bug in GHCi.
  If you suspect the latter, please send a bug report to:
  glasgow-haskell-bugs@haskell.org
ghcide: ^^ Could not load 'filepickerzmpolicyzm0zi2zi6zmLUnJDYkanFc1p7RlDh5zzSr_NetworkziFilepickerziHandle_FilepickerHandle_closure', dependency unresolved. See top entry above.

File:     entities/Model/User.hs
Range:    1:0-100001:0
Source:   typecheck
Severity: DsError
Message: 
  Program error:
  ByteCodeLink.lookupCE
  During interactive linking, GHCi couldn't find the following symbol:
  filepickerzmpolicyzm0zi2zi6zmLUnJDYkanFc1p7RlDh5zzSr_NetworkziFilepickerziHandle_FilepickerHandle_closure
  This may be due to you not asking GHCi to load extra object files,
  archives or DLLs needed by your current session. Restart GHCi, specifying
  the missing library using the -L/path/to/object/dir and -lmissinglibname
  flags, or simply by naming the relevant files on the GHCi command line.
  Alternatively, this link failure might indicate a bug in GHCi.
  If you suspect the latter, please send a bug report to:
  glasgow-haskell-bugs@haskell.org
Files that failed:
 * api/src/Daemon/Notification/Consignment.hs
```

My `hie.yaml` is:

```yaml
cradle:
  direct:
    arguments:
      - -Wall
      - -fno-warn-orphans
      - -hide-package resource-pool
      - -hide-package servant-checked-exceptions-core
      - -ich-aws/lib
      - -ich-eda/src
      - -ich-linear/lib
      - -ich-model/lib
      - -ich-persistence/src
      - -ich-segment/lib
      - -ich-servant-auth/lib
      - -ich-stripe/lib
      - -ich-wai-extra/lib
      - -idaemons/ch-shipment-tracking/src
      - -idaemons/ch-sync-mycronic-events/lib
      - -idaemons/ch-sync-trays/lib
      - -ieffects/hedis-simple-effects/lib
      - -ieffects/http-client-simple-effects/lib
      - -ientities
      - -iex-pool-prometheus/lib
      - -imouse-proxy/shared
      - -ioctopart
      - -ipicofactory/aoi/lib
      - -ipicofactory/coordinator/lib
      - -ipicofactory/jpctl/src
      - -ipicofactory/linescan-image-archiver/src
      - -ipicofactory/my300/src
      - -ipicofactory/panelisation/lib
      - -ipicofactory/ur10/lib
      - -ipicofactory/ur10/src-lib
      - -iprelude
      - -iquoting/lib
      - -package-env - # so that .ghci-environment.* generated by cabal is not used
```

The z-encoded name in question comes from something that is in my package database, as witnessed by:

```
$ ghc-pkg dump | grep Network.Filepicker.Handle
Network.Filepicker.Handle Network.Filepicker.Policy
```

Interestingly, even if I `-i` the source code of this module, I still get the same error - though I can see ghcide compiling the whole package as expected:

```
ghcide version: 0.0.5 (GHC: 8.8.1)
Ghcide setup tester in /home/ollie/work/circuithub.
Report bugs at https://github.com/digital-asset/ghcide/issues

Step 1/6: Finding files to test in /home/ollie/work/circuithub
Found 1 files

Step 2/6: Looking for hie.yaml files that control setup
Found 1 cradle

Step 3/6, Cradle 1/1: Loading ./hie.yaml

Step 4/6, Cradle 1/1: Loading GHC Session

Step 5/6: Initializing the IDE

Step 6/6: Type checking the files
File:     haskell-filepicker-policy/Network/Filepicker/Handle.hs
Range:    7:0-7:40
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/Handle.hs:7:1: warning:
  The qualified import of ‘Data.ByteString’ is redundant
  except perhaps to import instances from ‘Data.ByteString’
  To import instances alone, use: import Data.ByteString()
File:     haskell-filepicker-policy/Network/Filepicker/Policy.hs
Range:    22:0-22:48
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/Policy.hs:22:1: warning:
  The qualified import of ‘Data.Aeson.Types’ is redundant
  except perhaps to import instances from ‘Data.Aeson.Types’
  To import instances alone, use: import Data.Aeson.Types()
File:     haskell-filepicker-policy/Network/Filepicker/Policy.hs
Range:    23:0-23:53
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/Policy.hs:23:1: warning:
  The import of ‘Data.ByteArray’ is redundant
  except perhaps to import instances from ‘Data.ByteArray’
  To import instances alone, use: import Data.ByteArray()
File:     haskell-filepicker-policy/Network/Filepicker/Policy.hs
Range:    31:0-31:27
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/Policy.hs:31:1: warning:
  The import of ‘Data.Maybe’ is redundant
  except perhaps to import instances from ‘Data.Maybe’
  To import instances alone, use: import Data.Maybe()
File:     haskell-filepicker-policy/Network/Filepicker/Policy.hs
Range:    32:0-32:28
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/Policy.hs:32:1: warning:
  The import of ‘Data.Monoid’ is redundant
  except perhaps to import instances from ‘Data.Monoid’
  To import instances alone, use: import Data.Monoid()
File:     haskell-filepicker-policy/Network/Filepicker/Policy.hs
Range:    36:0-36:48
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/Policy.hs:36:1: warning:
  The qualified import of ‘Data.Text’ is redundant
  except perhaps to import instances from ‘Data.Text’
  To import instances alone, use: import Data.Text()
File:     haskell-filepicker-policy/Network/Filepicker/Policy.hs
Range:    43:0-43:28
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/Policy.hs:43:1: warning:
  The import of ‘Debug.Trace’ is redundant
  except perhaps to import instances from ‘Debug.Trace’
  To import instances alone, use: import Debug.Trace()
File:     haskell-filepicker-policy/Network/Filepicker/Policy.hs
Range:    71:0-71:10
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/Policy.hs:71:1: warning:
  Defined but not used: ‘setMaxSize’
File:     haskell-filepicker-policy/Network/Filepicker/Policy.hs
Range:    75:0-75:10
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/Policy.hs:75:1: warning:
  Defined but not used: ‘setMinSize’
File:     haskell-filepicker-policy/Network/Filepicker/Policy.hs
Range:    79:0-79:7
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/Policy.hs:79:1: warning:
  Defined but not used: ‘setPath’
File:     haskell-filepicker-policy/Network/Filepicker/Policy.hs
Range:    83:0-83:12
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/Policy.hs:83:1: warning:
  Defined but not used: ‘setContainer’
File:     haskell-filepicker-policy/Network/Filepicker/URL.hs
Range:    23:0-23:49
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/URL.hs:23:1: warning:
  The import of ‘Data.Monoid’ is redundant
  except perhaps to import instances from ‘Data.Monoid’
  To import instances alone, use: import Data.Monoid()
File:     haskell-filepicker-policy/Network/Filepicker/URL.hs
Range:    25:0-25:47
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/URL.hs:25:1: warning:
  The qualified import of ‘Data.Text’ is redundant
  except perhaps to import instances from ‘Data.Text’
  To import instances alone, use: import Data.Text()
File:     haskell-filepicker-policy/Network/Filepicker/URL.hs
Range:    54:14-54:20
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/URL.hs:54:15: warning:
  This binding for ‘handle’ shadows the existing binding
  imported from ‘Network.Filepicker.Policy’ at
  haskell-filepicker-policy/Network/Filepicker/URL.hs:28:1-42
  (and originally defined
  at haskell-filepicker-policy/Network/Filepicker/Policy.hs:50:5-10)
File:     haskell-filepicker-policy/Network/Filepicker/URL.hs
Range:    57:34-57:40
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/URL.hs:57:35: warning:
  This binding for ‘handle’ shadows the existing binding
  imported from ‘Network.Filepicker.Policy’ at
  haskell-filepicker-policy/Network/Filepicker/URL.hs:28:1-42
  (and originally defined
  at haskell-filepicker-policy/Network/Filepicker/Policy.hs:50:5-10)
File:     haskell-filepicker-policy/Network/Filepicker/URL.hs
Range:    64:26-64:32
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/URL.hs:64:27: warning:
  This binding for ‘handle’ shadows the existing binding
  imported from ‘Network.Filepicker.Policy’ at
  haskell-filepicker-policy/Network/Filepicker/URL.hs:28:1-42
  (and originally defined
  at haskell-filepicker-policy/Network/Filepicker/Policy.hs:50:5-10)
File:     haskell-filepicker-policy/Network/Filepicker/URL.hs
Range:    74:46-74:52
Source:   typecheck
Severity: DsWarning
Message: 
  haskell-filepicker-policy/Network/Filepicker/URL.hs:74:47: warning:
  This binding for ‘handle’ shadows the existing binding
  imported from ‘Network.Filepicker.Policy’ at
  haskell-filepicker-policy/Network/Filepicker/URL.hs:28:1-42
  (and originally defined
  at haskell-filepicker-policy/Network/Filepicker/Policy.hs:50:5-10)
ghcide: ^^ Could not load 'NetworkziFilepickerziHandle_FilepickerHandle_closure', dependency unresolved. See top entry above.

File:     entities/Model/Common.hs
Range:    1:0-100001:0
Source:   typecheck
Severity: DsError
Message: 
  Program error:
  ByteCodeLink.lookupCE
  During interactive linking, GHCi couldn't find the following symbol:
  NetworkziFilepickerziHandle_FilepickerHandle_closure
  This may be due to you not asking GHCi to load extra object files,
  archives or DLLs needed by your current session. Restart GHCi, specifying
  the missing library using the -L/path/to/object/dir and -lmissinglibname
  flags, or simply by naming the relevant files on the GHCi command line.
  Alternatively, this link failure might indicate a bug in GHCi.
  If you suspect the latter, please send a bug report to:
  glasgow-haskell-bugs@haskell.org
```

--- a/src/Development/IDE/Core/Compile.hs
+++ b/src/Development/IDE/Core/Compile.hs
@@ -47,9 +47,10 @@ import qualified GHC
 import           GhcMonad
 import           GhcPlugins                     as GHC hiding (fst3, (<>))
 import qualified HeaderInfo                     as Hdr
-import           HscMain                        (hscInteractive)
+import           HscMain                        (hscInteractive, hscSimplify)
 import           MkIface
 import           StringBuffer                   as SB
+import           TcRnMonad (tcg_th_coreplugins)
 import           TidyPgm
 
 import Control.Monad.Extra
@@ -148,9 +149,14 @@ compileModule packageState deps tmr =
                 let pm' = pm{pm_mod_summary = tweak $ pm_mod_summary pm}
                 let tm' = tm{tm_parsed_module  = pm'}
                 GHC.dm_core_module <$> GHC.desugarModule tm'
-
+            let tc_result = fst (tm_internals_ (tmrModule tmr))
+            -- Have to call the simplifier on the code even if we are at
+            -- -O0 as otherwise the code generation fails which leads to
+            -- errors like #256
+            plugins <- liftIO $ readIORef (tcg_th_coreplugins tc_result)
+            desugared_guts <- liftIO $ hscSimplify session plugins desugar
             -- give variables unique OccNames
-            (guts, details) <- liftIO $ tidyProgram session desugar
+            (guts, details) <- liftIO $ tidyProgram session desugared_guts
             return (map snd warnings, (mg_safe_haskell desugar, guts, details))
 
 generateByteCode :: HscEnv -> [TcModuleResult] -> TcModuleResult -> CgGuts -> IO (IdeResult Linkable)

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -1374,6 +1374,27 @@ thTests =
         _ <- openDoc' "A.hs" "haskell" sourceA
         _ <- openDoc' "B.hs" "haskell" sourceB
         expectDiagnostics [ ( "B.hs", [(DsError, (6, 29), "Variable not in scope: n")] ) ]
+    , testSessionWait "newtype-closure" $ do
+        let sourceA =
+              T.unlines
+                [ "{-# LANGUAGE DeriveDataTypeable #-}"
+                  ,"{-# LANGUAGE TemplateHaskell #-}"
+                  ,"module A (a) where"
+                  ,"import Data.Data"
+                  ,"import Language.Haskell.TH"
+                  ,"newtype A = A () deriving (Data)"
+                  ,"a :: ExpQ"
+                  ,"a = [| 0 |]"]
+        let sourceB =
+              T.unlines
+                [ "{-# LANGUAGE TemplateHaskell #-}"
+                ,"module B where"
+                ,"import A"
+                ,"b :: Int"
+                ,"b = $( a )" ]
+        _ <- openDoc' "A.hs" "haskell" sourceA
+        _ <- openDoc' "B.hs" "haskell" sourceB
+        return ()
     ]
 
 completionTests :: TestTree

